text/0019-fork-spam.md

```diff
@@ -0,0 +1,62 @@
+[2]: https://github.com/xcthulhu/casper-node/blob/eae7c437819c1ee0f4f33e4373ff66fb5e98379a/node/src/components/consensus/protocols/highway.rs#L379-L442
+
+## Motivation
+
+[motivation]: #motivation
+
+This CEP addresses a potential security threat.  The threat is a *Denial of Service* (DOS) attack. An equivocating node could spam the network with many faulty votes.  These would result in the node producing block with a large number of dependencies in the current implementation.  These blocks with spam dependencies could stop the network from processing nodes. 
+
+## Guide-level explanation
+
+[guide-level-explanation]: #guide-level-explanation
+
+*Fork spam* refers to when an equivocating node broadcasts many votes over the wire to its peers.
+
+Properly functioning nodes should not include this vote spam in the protocol state.  However, an exception is made if a spammed vote is referenced by another properly functioning node's vote.
+
+The following UML diagram shows how an attacking equivocator will spam a node, resulting in a very large protocol state with 10000000000 votes.
+
+![Current Implementation Vulnerable To Fork Spam](images/0019/fork_spam.svg)
+
+After our proposed changes, an equivocator could at most once get a spammed vote included in the protocol state.  They would have to deceive an honest node.  Here is an UML diagram showing how messages will be processed:
+
+![Current Implementation Vulnerable To Fork Spam](images/0019/fork_spam_new.svg)
+
+For implementation-oriented CEPs (e.g. for node internals), this section should focus on how other developers should think about the change, and give examples of its concrete impact. For policy CEPs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Votes are to be annotated with equivocators if detected in [`components::consensus::highway_core::highway::Highway::pre_validate_vertex`][1]. Then, in [`components::consensus::protocols::highway::HighwayProtocol::handle_message`][2] votes are processed as follows:
+
+  1. If the vote is not from an equivocator, process as normal
+  2. If it's from an equivocator and from the future, discard
+  3. If it's from an equivocator and a (transitive) dependency of another vote, keep it
+  4. Otherwise discard it
+
+(3) captures the workflow with the deceived honest validator presented in the last section.
+
+## Rationale and alternatives
+
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+An alternate design would be to combine `components::consensus::highway_core::highway::Highway` and `components::consensus::protocols::highway::HighwayProtocol`.  The scope of this would be rather large.
+
+## Unresolved questions
+
+[unresolved-questions]: #unresolved-questions
+
+- Do we need to disconnect spamming equivocators using the gossip network?
```

Comment:
I agree we should but it might be tricky to not disconnect from honest nodes that only forward the equivocations.  We're not only exchanging votes as a dependency (Alice sends her vote `a0` to Bob, Bob creates a message `b0` citing `a0` and sends it to Charlie), but we also (or will do soon) gossip (Alice sends her vote `a0` to Bob, Bob receives it and gossips to Charlie, Ed, Frank, etc.).

For example, Alice equivocates with A0, A1, A2. She sends A0 and A1 to Bob, and A1 and A2 to Charlie. Bob will gossip both votes to Charlie. Charlie recognizes that A2 is an excessive vote created by Alice The Equivocator and decides to ban Bob. 

This is not captured in the UML diagrams, am I right?

Reply:
@xcthulhu created [HWY-187](https://casperlabs.atlassian.net/browse/HWY-187) for that. I guess that will be a design task in itself. I agree that we mustn't have false positives here.

Reply:
Ok, my understanding was that is an implementation task. *This* CEP though is about defence against fork spam attacks so I thought it would make sense to have everything, that relates to it, in one place.

Reply:
Right; let's clarify that in today's standup.

Reply:
@goral09 I've gone ahead and implemented the non-gossip network bit of this in my PR here: https://github.com/CasperLabs/casper-node/pull/516

I would like to tackle the design for [HWY-187][1] in another CEP.  Does that work with you?  Keeping this design doc less than 500 words might make this easier for @fizyk20 to review if he has the cycles...

[1]: https://casperlabs.atlassian.net/browse/HWY-187

Reply:
Sure, let's do that.